### PR TITLE
RYA-443 Rya Streams Query Manager daemon program.

### PR DESCRIPTION
**Pull request by kchilton2**
## Description
Implemented an RPM installed CentOS 7 daemon service used to execute Kafka Streams jobs on the local machine. This is part of Rya Streams.

### Tests
Wrote automated tests and executed the application on a CentOS 7 machine using the RPM.

### Links
https://issues.apache.org/jira/browse/RYA-443

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/686/<h2>Build result: FAILURE</span></h2>[...truncated 3.04 MB...][INFO] Apache Rya Web Projects ............................ SKIPPED[INFO] Apache Rya Web Implementation ...................... SKIPPED[INFO] ------------------------------------------------------------------------[INFO] BUILD FAILURE[INFO] ------------------------------------------------------------------------[INFO] Total time: 38:10 min[INFO] Finished at: 2018-02-08T22:18:44+00:00[INFO] Final Memory: 301M/2452M[INFO] ------------------------------------------------------------------------Waiting for Jenkins to finish collecting data[ERROR] Failed to execute goal org.codehaus.mojo:rpm-maven-plugin:2.1.5:attached-rpm (create-rpm-distribution) on project rya.streams.query-manager: Unable to query for default vendor from RPM: Error while executing process. Cannot run program "rpm": error=2, No such file or directory -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :rya.streams.query-managerchannel stoppedSetting status of b889bc0d60e3884ea31edbcb12fcc21331f77963 to FAILURE with url https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/686/ and message: 'FAILURE 'Using context: Jenkins: clean package -Pgeoindexing

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/687/<h2>Build result: FAILURE</span></h2>[...truncated 3.01 MB...][INFO] Apache Rya Spark Support ........................... SKIPPED[INFO] Apache Rya Web Projects ............................ SKIPPED[INFO] Apache Rya Web Implementation ...................... SKIPPED[INFO] ------------------------------------------------------------------------[INFO] BUILD FAILURE[INFO] ------------------------------------------------------------------------[INFO] Total time: 40:57 min[INFO] Finished at: 2018-02-08T22:25:43+00:00[INFO] Final Memory: 546M/3106M[INFO] ------------------------------------------------------------------------[ERROR] Failed to execute goal org.codehaus.mojo:rpm-maven-plugin:2.1.5:attached-rpm (create-rpm-distribution) on project rya.streams.query-manager: Unable to query for default vendor from RPM: Error while executing process. Cannot run program "rpm": error=2, No such file or directory -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :rya.streams.query-managerchannel stoppedSetting status of 50a81217e5b16a4b1377f199aefc3a1c5a225216 to FAILURE with url https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/687/ and message: 'FAILURE 'Using context: Jenkins: clean package -Pgeoindexing

File: extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java [hunk 1]
```diff
@@ -107,12 +112,11 @@ private String formatQueries(final Set<StreamsQuery> queries) {
...
-            sb.append("ID: ");
-            sb.append(query.getQueryId());
-            sb.append("\t\t");
-            sb.append("Query: ");
-            sb.append(query.getSparql());
-            sb.append("\n");
+            sb.append("ID: ").append(query.getQueryId())
+                .append("    ")
+                .append("Is Active: ").append(query.isActive())
+                .append( query.isActive() ? "     " : "    " )
```
Comment (ejwhite922) [inline extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java]: Maybe change to:

<code>
.append("Is Active: ")

.append(StringUtils.rightPad(query.isActive(), 9))
</code>

File: extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/entity/StreamsQuery.java [hunk 2]
```diff
@@ -85,4 +85,19 @@ public boolean equals(final Object o) {
...
         return false;
     }
+    
+    @Override
+    public String toString() {
+        final StringBuilder sb = new StringBuilder();
+        sb.append("ID: ");
+        sb.append(getQueryId().toString() + "\n");
+        sb.append("Query: ");
+        sb.append(getSparql() + "\n");
```
Comment (ejwhite922) [inline extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/entity/StreamsQuery.java]: Is "getSparql()" coming from user input? Might want to wrap that in org.apache.rya.api.log.LogUtils.clean(getSparql())

File: extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java [hunk 3]
```diff
@@ -107,12 +112,11 @@ private String formatQueries(final Set<StreamsQuery> queries) {
...
+                .append("Query: ").append(query.getSparql()).append("\n");
```
Comment (ejwhite922) [inline extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java]: LogUtils.clean(query.getSparql())

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/kafka/KafkaQueryChangeLogSource.java [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.AbstractScheduledService;
+
+import edu.umd.cs.findbugs.annotations.DefaultAnnotation;
+import edu.umd.cs.findbugs.annotations.NonNull;
+
+/**
+ * Represents a Kafka Server that manages {@link KafkaQueryChangeLog}s.
+ * <p/>
+ * Thread safe.
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/kafka/KafkaQueryChangeLogSource.java]: What's the dependency for that annotation?

**Comment by kchilton2** [comment]
How do we want to deal with this project producing an RPM? The ASF jenkins doesn't have the rpm command installed.

File: extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/entity/StreamsQuery.java [hunk 5]
```diff
@@ -85,4 +85,19 @@ public boolean equals(final Object o) {
...
+        sb.append("Is ");
+        if (!isActive) {
+            sb.append(" Not ");
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/entity/StreamsQuery.java]: Done.

File: extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/queries/InMemoryQueryRepository.java [hunk 6]
```diff
@@ -243,8 +274,68 @@ private void updateCache() {
...
+    @Override
+    protected void runOneIteration() throws Exception {
+        log.trace("runOneIteration() - Enter");
+        lock.lock();
+        try {
+            updateCache();
+        } finally {
+            lock.unlock();
+        }
+        log.trace("runOneIteration() - Exit");
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/queries/InMemoryQueryRepository.java]: Yep, done.

File: extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/queries/QueryChange.java [hunk 7]
```diff
@@ -110,6 +110,16 @@ public boolean equals(final Object o) {
         return false;
     }
 
+    @Override
+    public String toString() {
+        return "QueryChange: {" +
+               "    Query ID: " + queryId + ",\n" +
+               "    Change Type: " + changeType + ",\n" +
+               "    Is Active: " + isActive + ",\n" +
+               "    SPARQL: " + sparql + "\n" +
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/queries/QueryChange.java]: I'm going to assume it is sanitized by the time it reaches this line of code.

File: extras/rya.streams/api/src/test/java/org/apache/rya/streams/api/queries/InMemoryQueryRepositoryTest.java [hunk 8]
```diff
@@ -85,11 +92,16 @@ public void initializedWithPopulatedChangeLog() throws Exception {
...
 
             // Create a new totally in memory QueryRepository.
-            try(final QueryRepository initializedQueries = new InMemoryQueryRepository( changeLog )) {
+            final QueryRepository initializedQueries = new InMemoryQueryRepository( changeLog, SCHEDULE );
+            try {
                 // Listing the queries should work using an initialized change log.
                 final Set<StreamsQuery> stored = initializedQueries.list();
                 assertEquals(expected, stored);
+            } finally {
+                queries.stop();
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/api/src/test/java/org/apache/rya/streams/api/queries/InMemoryQueryRepositoryTest.java]: Removed.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/KafkaTopics.java [hunk 9]
```diff
@@ -48,7 +51,28 @@
      */
     public static String queryChangeLogTopic(final String ryaInstance) {
         requireNonNull(ryaInstance);
-        return ryaInstance + "-QueryChangeLog";
+        return ryaInstance + QUERY_CHANGE_LOG_TOPIC_SUFFIX;
+    }
+
+    /**
+     * Get the Rya instance name from a Kafka topic name that has been used for a {@link QueryChnageLog}.
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/KafkaTopics.java]: Done.

File: extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/SingleThreadKafkaStreamsFactory.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        streamsProps.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+
+        // Setup the topology that processes the Query.
+        final String statementsTopic = KafkaTopics.statementsTopic(ryaInstance);
+        final String resultsTopic = KafkaTopics.queryResultsTopic(query.getQueryId());
+
+        try {
+            final TopologyBuilder topologyBuilder = topologyFactory.build(query.getSparql(), statementsTopic, resultsTopic, new RandomUUIDFactory());
+            return new KafkaStreams(topologyBuilder, new StreamsConfig(streamsProps));
+        } catch (MalformedQueryException | TopologyBuilderException e) {
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/kafka/src/main/java/org/apache/rya/streams/kafka/SingleThreadKafkaStreamsFactory.java]: Done.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java [hunk 11]
```diff
@@ -0,0 +1,884 @@
...
+     * @param offerValue - How long to wait when offering new work.
+     * @param offerUnits - The unit for the {@code offerValue}. (not null)
+     * @param shutdownSignal - Used to signal application shutdown has started, so
+     *   this method may terminate without ever placing the event on the queue. (not null)
+     * @return {@code true} if the evet nwas added to the queue, otherwise false.
+     */
+    private static <T> boolean offerUntilAcceptedOrShutdown(
+            final BlockingQueue<T> workQueue,
+            final T event, final
+            long offerValue,
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java]: Done.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java [hunk 12]
```diff
@@ -0,0 +1,884 @@
...
+        /**
+         * Make a {@link LogEvent} that indicates a {@link QueryChangeLog} was created within a
+         * {@link QueryChangeLogSource}.
+         *
+         * @param ryaInstance - The Rya Instance the created log is for. (not null)
+         * @param log - The created {@link QueryChangeLog. (not null)
+         * @return A {@link LogEvent} built using the provided values.
+         */
+        public static LogEvent create(final String ryaInstance, final QueryChangeLog log) {
+            return new LogEvent(ryaInstance, LogEventType.CREATE ,Optional.of(log));
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java]: Done.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java [hunk 13]
```diff
@@ -0,0 +1,884 @@
...
+                        Objects.equals(query, other.query);
+            }
+            return false;
+        }
+
+        @Override
+        public String toString() {
+            final StringBuilder string = new StringBuilder();
+            string.append("Query Event {\n")
+                  .append("    Rya Instance:").append(ryaInstance).append(",\n")
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java]: Done.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java [hunk 14]
```diff
@@ -0,0 +1,884 @@
...
+            return string.toString();
+        }
+
+        private void append(final StringBuilder string, final StreamsQuery query) {
+            requireNonNull(string);
+            requireNonNull(query);
+            string.append("    Streams Query {\n")
+                  .append("        Query ID: ").append(query.getQueryId()).append(",\n")
+                  .append("        Is Active: ").append(query.isActive()).append(",\n")
+                  .append("        SPARQL: ").append(query.getSparql()).append("\n")
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java]: I'm going to assume it has been sanitized by this point.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java [hunk 15]
```diff
@@ -0,0 +1,884 @@
...
+        private final BlockingQueue<QueryEvent> workQueue;
+        private final QueryExecutor queryExecutor;
+        private final long pollingValue;
+        private final TimeUnit pollingUnits;
+        private final AtomicBoolean shutdownSignal;
+
+        /**
+         * Constructs an instance of {@link QueryEventWorker}.
+         *
+         * @param logWorkQueue - A queue of {@link QueryEvent}s that will be worked by this object. (not null)
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java]: Done.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java [hunk 16]
```diff
@@ -0,0 +1,884 @@
...
+                        case STOPPED:
+                            try {
+                                queryExecutor.stopQuery(event.getQueryId().get());
+                            } catch (final IllegalStateException | QueryExecutorException e) {
+                                log.error("Could not stop a query represented by the following work: " + event, e);
+                            }
+                            break;
+
+                        case STOP_ALL:
+                        try {
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManager.java]: Done.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManagerDaemon.java [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+            new JCommander(params).parse(args);
+        } catch(final ParameterException e) {
+            throw new DaemonInitException("Unable to parse the command line arguments.", e);
+        }
+        final Path configFile = params.config != null ? Paths.get(params.config) : DEFAULT_CONFIGURATION_PATH;
+        log.info("Loading the following configuration file: " + configFile);
+
+        // Unmarshall the configuration file into an object.
+        final QueryManagerConfig config;
+        try(InputStream stream = Files.newInputStream(configFile)) {
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/QueryManagerDaemon.java]: Done.

File: extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/ThreadUtil.java [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+ * under the License.
+ */
+package org.apache.rya.streams.querymanager;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Utilities that are useful for interacting with {@link Thread}s while testing.
+ */
+public class ThreadUtil {
```
[the remark this answers is not in the document]
Comment (kchilton2) [reply extras/rya.streams/query-manager/src/test/java/org/apache/rya/streams/querymanager/ThreadUtil.java]: Done.

File: extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java (hunk 1 above)
Comment (kchilton2) [reply extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java]: Done.

File: extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/entity/StreamsQuery.java (hunk 2 above)
Comment (kchilton2) [reply extras/rya.streams/api/src/main/java/org/apache/rya/streams/api/entity/StreamsQuery.java]: Nope.

File: extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java (hunk 3 above)
Comment (kchilton2) [reply extras/rya.streams/client/src/main/java/org/apache/rya/streams/client/command/ListQueriesCommand.java]: Should b e sanitized before it gets here.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/691/<h2>Build result: FAILURE</span></h2>[...truncated 3.03 MB...][INFO] Apache Rya Spark Support ........................... SKIPPED[INFO] Apache Rya Web Projects ............................ SKIPPED[INFO] Apache Rya Web Implementation ...................... SKIPPED[INFO] ------------------------------------------------------------------------[INFO] BUILD FAILURE[INFO] ------------------------------------------------------------------------[INFO] Total time: 37:36 min[INFO] Finished at: 2018-02-14T20:56:07+00:00[INFO] Final Memory: 535M/2904M[INFO] ------------------------------------------------------------------------[ERROR] Failed to execute goal org.codehaus.mojo:rpm-maven-plugin:2.1.5:attached-rpm (create-rpm-distribution) on project rya.streams.query-manager: Unable to query for default vendor from RPM: Error while executing process. Cannot run program "rpm": error=2, No such file or directory -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :rya.streams.query-managerchannel stoppedSetting status of 6428ca81a5cb0ef5bd2e762452f1a6e46dff88fc to FAILURE with url https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/691/ and message: 'FAILURE 'Using context: Jenkins: clean package -Pgeoindexing

File: extras/rya.streams/query-manager/pom.xml [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+                <artifactId>rpm-maven-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <id>set-rpm-properties</id>
+                        <goals>
+                            <goal>version</goal>
+                        </goals>
+                    </execution>
+                    <execution>
+                        <id>create-rpm-distribution</id>
```
[the remark this answers is not in the document]
Comment (ejwhite922) [reply extras/rya.streams/query-manager/pom.xml]: Probably both but not needed for this PR.

File: extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/kafka/KafkaQueryChangeLogSource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (ejwhite922) [reply extras/rya.streams/query-manager/src/main/java/org/apache/rya/streams/querymanager/kafka/KafkaQueryChangeLogSource.java]: net.jcip.annotations.ThreadSafe

Looks like Rya is using this one:
```
        <dependency>
            <groupId>com.github.stephenc.jcip</groupId>
            <artifactId>jcip-annotations</artifactId>
            <version>1.0-1</version>
        </dependency>
```

**Comment by kchilton2** [comment]
I'm closing this because https://github.com/apache/incubator-rya/pull/279 takes it over.

**Pull request closed without merging**